docs/USAGE-GUIDE.md

```diff
@@ -654,6 +654,18 @@ If the read end of the pipe is closed unexpectedly, writing to the pipe will rai
 a SIGPIPE signal. **s2n-tls does NOT handle SIGPIPE.** A SIGPIPE signal will cause
 the process to terminate unless it is handled or ignored by the application.
 
+**WARNING (TLS v1.2 ONLY):** Enabling Nagle's algorithm can increase handshake latency
+* Enabling Nagle's algorithm can increase the handshake latency. To avoid this behaviour, please disable Nagle's at the socket level; this can be done with the snippet displayed below. Please note this is only applicable to TLS v1.2.
+* If enabling Nagle's is required, then an increase in handshake latency is expected. This behavior occurs because `s2nc` does not coalesce multiple handshake messages into a record. In effect, after the `s2nc` sends the client key exchange record to the server, it waits until it receives a server ack before continuing the handshake.
```

Comment:
Did we confirm this was the root cause?

Also, it's not s2nc that doesn't bundle multiple messages into single records, it's s2n_negotiate. s2nc is just an example application.

Reply:
Yes, I confirmed it w/ Panos. I'll change to `s2n_negotiate` or just `s2n`.

Reply:
If that's the case, and we're confident enough to add this to our documentation, can you explain how enabling Nagle's leads to slower handshakes? What happens with Nagle's vs without Nagle's? How does it interact with corking?

If we're not confident, let's hold off on adding this warning.

Reply:
(You don't have to put that info in the docs, but let's at least document it in this PR)

Reply:
I linked an rfc describing Nagle's in the change (https://www.rfc-editor.org/rfc/rfc896). In a nutshell, Nagle's algorithm is if the amount of data is smaller than some limit then delay the send until an ACK is received (for some previously sent data). Otherwise, send the packet. 

Corking is similar but it does not delay sending until an ACK is received, but it delays sending until the buffer is filled to some limit. So this is why we observe the extra latency when Nagle's is enabled. Because s2n is sending multiple records, Nagle's will eventually delay sending data until it receives an ack. In Panos and my wireshark traces, we see this happening between the client key exchange and change cipher spec messages.

Reply:
But if we're using corking, and corking means we delay until a buffer is filled (or the socket is uncorked, which we should do at the end of every message flight) instead of until an ACK is received, why would this be a problem? Shouldn't we "flush" at the end of every flight?

Reply:
And since that's our intended behavior, if Nagle's interferes with that behavior, shouldn't we disable it as part of our corking logic?

Reply:
Not sure if I follow 100%, but I think what you're saying is that if corking is configured, then we shouldn't run into the issue mentioned in this pr. In that case, it's true but I think that would also depend on the buffer limit + the size of the records. In this case, I did observe that with corking on then we don't have the issue mentioned above (attached as a wireshark trace with corking & nagle's on). However, this issue is present with corking off (since we're not buffering messages to some limit).

I don't think we need to disable Nagle's when corking is set. They can be configured independently of each other
<img width="1486" alt="nagle_on_cork_on" src="https://user-images.githubusercontent.com/29711666/185220654-fffebf26-d064-45c9-a747-63e21df15a11.png">
.

Reply:
Update: So I did some manual runs and I observed the following:

| Corking enabled? | Nagle's enabled? | No issues observed? |
|------------------|------------------|---------------------|
| ✅                | ❌                | ✅                   |
| ✅                | ✅                | ❌                   |
| ❌                | ✅                | ❌                   |

In summary: looks like Nagle's + Corking still exhibits the delayed-handshake issue. We can either note this in the docs or we can proactively disable Nagel's while s2n is corking.